src/librustc/middle/ty.rs

```diff
@@ -2301,6 +2289,128 @@ impl<'a, 'tcx> ParameterEnvironment<'a, 'tcx> {
     }
 }
 
+pub struct DatatypeDef<'tcx> {
+    pub def_id: DefId,
+    pub variants: Vec<VariantDef<'tcx>>,
+}
+
+pub struct VariantDef<'tcx> {
+    pub id: DefId,
+    pub name: Name,
+    pub disr_val: Disr,
+    pub fields: Vec<FieldTy<'tcx>>,
+    pub vis: Visibility,
+}
+
+
+pub struct FieldTy<'tcx> {
+    pub id: DefId,
+    pub name: Name,
+    pub vis: ast::Visibility,
+    pub origin: DefId,
+    pub ty: Cell<Ty<'tcx>>
```

Comment:
This needs a comment explaining why `Cell` is used here.
Similarly, the distinction between `id` and `origin` is not obvious, could also use a comment.